runtime/cudaq/dynamics/cudm_state.cpp

```diff
@@ -0,0 +1,252 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include <cmath>
+#include <cudaq/cudm_state.h>
+#include <iostream>
+#include <numeric>
+#include <sstream>
+#include <stdexcept>
+
+namespace cudaq {
+
+cudm_state::cudm_state(std::vector<std::complex<double>> rawData)
+    : rawData_(rawData), state_(nullptr), handle_(nullptr),
+      hilbertSpaceDims_() {
+  HANDLE_CUDM_ERROR(cudensitymatCreate(&handle_));
+}
+
+cudm_state::~cudm_state() {
+  if (state_) {
+    cudensitymatDestroyState(state_);
+  }
+  if (handle_) {
+    cudensitymatDestroy(handle_);
+  }
```

Comment:
Like above, we shouldn't be creating and destroying the handle during state object life cycle.